### PR TITLE
Podman machine info

### PR DESCRIPTION
**Pull request by ashley-cui**
Add podman machine info command, which displays info about the machine
host as well as version info.

Signed-off-by: Ashley Cui <acui@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
Add podman machine info command, which displays info about the machine
host as well as version info.

```

### CODE REVIEW

**Comment by ashley-cui** [comment]
Closes: https://github.com/containers/podman/issues/14113

File: cmd/podman/machine/info.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+	provider := GetSystemDefaultProvider()
+	var listOpts machine.ListOptions
+	listResponse, err := provider.List(listOpts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get machines")
+	}
+
+	host.NumberOfMachines = len(listResponse)
+	host.RemoteSocket = fmt.Sprintf("/run/user/%d/podman/podman.sock", os.Getuid())
```
Comment (ashley-cui) [inline cmd/podman/machine/info.go:130]: This is the only field I'm not sure is entirely correct, please correct me if I'm wrong here.

**Comment by ashley-cui** [comment]
@baude @benoitf PTAL

File: cmd/podman/machine/info.go (hunk 1 above)
Comment (cdoern) [inline cmd/podman/machine/info.go:130]: if running rootful do we need to check for `/run/user/podman.sock`?

**Review by cdoern — COMMENTED** [review]
Is there any info to indicate whether the machine is starting? Or would that not be pertinent here, only in machine list?

File: cmd/podman/machine/info.go (hunk 1 above)
Comment (ashley-cui) [reply cmd/podman/machine/info.go:130]: Wouldn't that just be the same as eventSockDIr? Or is remote socket something different? @baude, thoughts?

**Comment by benoitf** [comment]
That command is really helpful.

Few notes:
- it has been said by someone else but yes I would expect to know if a Podman machine is currently starting (or stopping)
- I would expect things changing to be prefixed by 'Current' maybe (or be in a sub-section)
  For example `RemoteSocket` will change depending on the machine we start (and the same for the currently running machine `RunningMachine`) while `arch`, `DefaultMachine`, `EventsDir`, etc. are for all machines
  
  like:
  ```json
  "Host": {
  "NumberOfMachines": 1, 
   "current" : {
      "Name": "podman-machine-default",
      "Starting": true,
      "Running": false,
      "Stopping": false,
      "RemoteSocket": "/run/user/501/podman/podman.sock"
}
  ```

File: pkg/machine/e2e/info_test.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	)
+
+	BeforeEach(func() {
+		testDir, mb = setup()
+	})
+	AfterEach(func() {
+		teardown(originalHomeDir, testDir, mb)
+	})
+
+	It("machine info", func() {
```
Comment (baude) [inline pkg/machine/e2e/info_test.go:24]: could you sprinkle in comments on tests like this where you are looking at a lot of different things?

File: cmd/podman/machine/info.go (hunk 1 above)
Comment (baude) [reply cmd/podman/machine/info.go:130]: It is defined right here for podman info -> https://github.com/containers/podman/blob/6c4c050d3df4f4e952273891bddfeb5b5bd8be5e/pkg/domain/infra/abi/system.go#L46

**Comment by baude** [comment]
would it be enough to list the machines by name and then use the statuses to determine which is "current"?

**Comment by mheon** [comment]
Why are we including a list of machines? That seems like a `podman machine ls` thing?

**Comment by benoitf** [comment]
I would like to have the 'LocalSocket' as well

because for example `RemoteSocket: /run/user/501/podman/podman.sock` is not so interesting while using a machine

I would prefer to know the `/Users/benoitf/.local/share/containers/podman/machine/podman-machine-default/podman.sock` path

**Comment by baude** [comment]
@mheon thanks for the reminder on machine ls; indeed, I did not want to list each machine in machine info; that is what `podman machine ls` is for ... i think the original idea was to give number of machines; and if a machine was running, to show that.

**Comment by mheon** [comment]
Number of machines/number of running, maybe? Knowing at least 1 is up is a clue to go to `machine ls` to check, I'd imagine.

**Comment by ashley-cui** [comment]
@mheon, @baude , @benoitf  I'd agree that we shouldn't list machines in machine info, and I'd also prefer if we don't add any information on the machine, as that's covered by `podman machine inspect`. `podman machine info` should be generic, ie the information in podman machine info should be relevant to every machine on the system.  I think if we do display the machine socket, that should be in inspect. 

@mheon, Only one machine can be up at a time, so maybe a field like machineRunning that's a bool?

**Comment by benoitf** [comment]
I'm ok to remove all the content that is related to the current machine (like sockets, etc) (and then add it in the inspect command)

About the default machine, can we change through cli the default machine name or is it always 'podman-machine-default' ?

but I think user wants to know the current status of the machine: running, starting or stopped (Running is not enough)

because if you do `podman machine info` and the machine is still not displayed as `RunningMachine` you may try to execute `podman machine start` while the machine is already trying to start

I would also be interested to know if machine has failed to start (it's not running but there was a failure)

so something like

```json
  "CurrentMachine": "podman-machine-default",
  "State|Status": "Starting|Running|Failed",
```

**Comment by ashley-cui** [comment]
Sure, I think default machine is the same thing as current machine. I think I would be okay with adding something like default machine state too, if the others are as well. :smile:

**Comment by baude** [comment]
i assume current machine to mean -- the machine that is running and has nothing to do with default.  And using that definition, if a machine is running, the state must be `running` as well.  Am I viewing this differently than others?  @rhatdan what so you?

**Comment by rhatdan** [comment]
Since we only have one machine running at a time, I would expect it to be current machine, if no machine is running then it should be default.

**Comment by ashley-cui** [comment]
Updated!

**Comment by rhatdan** [comment]
LGTM
@benoitf @baude PTAL

**Comment by benoitf** [comment]
I tried again the command

when I launch podman machine info without a machine I can see
```
Host:
  Arch: amd64
  CurrentMachine: ""
  DefaultMachine: ""
  EventsDir: /var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/podman-run--1/podman
  MachineConfigDir: /Users/benoitf/.config/containers/podman/machine/qemu
  MachineImageDir: /Users/benoitf/.local/share/containers/podman/machine/qemu
  MachineState: ""
  NumberOfMachines: 0
  OS: darwin
  RemoteSocket: /var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/podman-run--1/podman/podman.sock
  VMType: qemu
```

But the RemoteSocket path does not exists
```bash
$ ls -la /var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/podman-run--1/podman/podman.sock
ls: /var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/podman-run--1/podman/podman.sock: No such file or directory
```

on my side the socket is at
`~/.local/share/containers/podman/machine/podman.sock` symlink to `/Users/benoitf/.local/share/containers/podman/machine/podman-machine-default/podman.sock` 



If I only create a machine but I don't start it
`podman machine info` reports

```
RemoteSocket: /var/folders/tg/_5rxbhmj4xncz4szvpgswrmc0000gn/T/podman-run--1/podman/podman.sock
```
but socket path still doesn't exists

and even if I do run the machine the path is still incorrect ?

so
 1. the path displayed is invalid ?
 2. I don't know if RemoteSocket should be displayed if there is no machine or if there is no machine running


as further improvement, I would like to see, if possible, that a machine is being created/initialized

if I do `podman machine init` in one terminal and `podman machine info` in another one

I would expect to have when doing `podman machine info`

```
  ...
  CurrentMachine: podman-machine-default
  DefaultMachine: podman-machine-default
  ...
  MachineState: Creating
  NumberOfMachines: 1
  ```

**Comment by ashley-cui** [comment]
Hmm, okay I think then it isn't XDG_RUNTIME_DIR, let me take another look and fix it

**Comment by ashley-cui** [comment]
Okay, socket should be correct, but I'm not sure if we should put the sock in machine info, since now we're going into machine inspect territory, by calling inspect. WDYT @baude

File: docs/source/markdown/podman-machine-info.1.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+#### **--help**
+
+Print usage statement.
+
+## EXAMPLES
+
+```
+$ podman machine info
+$ podman machine info --format json
```
Comment (TomSweeneyRedHat) [inline docs/source/markdown/podman-machine-info.1.md:28]: might be nice to have a simple go template example too.

**Comment by baude** [comment]
i think whether the socket exists or not is outside the scope of this PR.  I would recommend that we merge this PR and have @benoitf create an issue about the sockets.  One thing regarding the sockets, we intentionally do not create them unless a machine is running but for machine events, it probably does make sense.

**Comment by baude** [comment]
LGTM

File: docs/source/markdown/podman-machine-info.1.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+## NAME
+podman\-machine\-info - Display machine host info
+
+## SYNOPSIS
+**podman machine info**
+
+## DESCRIPTION
+
+Display information pertaining to the machine host.
+Rootless only.
```
Comment (mheon) [inline docs/source/markdown/podman-machine-info.1.md:12]: Can you elaborate on this? "Can only be used with rootless Podman as `podman machine` can only be used with rootless Podman" or similar?

File: pkg/machine/e2e/config_info.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package e2e
+
+type infoMachine struct {
+	/*
+		--format string   Format volume output using JSON or a Go template)
```
Comment (mheon) [inline pkg/machine/e2e/config_info.go:5]: Can this be removed? Doesn't really seem related to the struct

**Comment by mheon** [comment]
Two nits otherwise LGTM

File: docs/source/markdown/podman-machine-info.1.md (hunk 4 above)
Comment (mheon) [reply docs/source/markdown/podman-machine-info.1.md]: Sure

**Comment by ashley-cui** [comment]
Tests are green, @containers/podman-maintainers PTAL and merge

**Pull request merged**